grpc-api/src/main/java/io/deephaven/grpc_api/console/ConsoleServiceGrpcImpl.java

```diff
@@ -253,106 +255,91 @@ public void fetchTable(FetchTableRequest request, StreamObserver<ExportedTableCr
         });
     }
 
-    // TODO(core#101) autocomplete support
-    @Override
-    public void openDocument(OpenDocumentRequest request, StreamObserver<OpenDocumentResponse> responseObserver) {
-        // when we open a document, we should start a parsing thread that will monitor for changes, and pre-parse document
-        // so we can respond appropriately when client wants completions.
-        GrpcUtil.rpcWrapper(log, responseObserver, () -> {
-            final SessionState session = sessionService.getCurrentSession();
-            SessionState.ExportObject<ScriptSession> exportedConsole = session.getExport(request.getConsoleId());
-            session
-                    .nonExport()
-                    .require(exportedConsole)
-                    .onError(responseObserver::onError)
-                    .submit(()->{
-                        final ScriptSession scriptSession = exportedConsole.get();
-                        final TextDocumentItem doc = request.getTextDocument();
-                        scriptSession.getParser().open(doc.getText(), doc.getUri(), Integer.toString(doc.getVersion()));
-                        safelyExecute(() -> {
-                            responseObserver.onNext(OpenDocumentResponse.getDefaultInstance());
-                            responseObserver.onCompleted();
-                        });
-                    });
+    private CompletionParser ensureParserForSession(SessionState session) {
+        return parsers.computeIfAbsent(session, s -> {
+            CompletionParser parser = new CompletionParser();
+            s.addOnCloseCallback(parser);
```

Comment:
This on close callback also needs to remove from the parsers map.

Reply:
No, this is deliberate (and I'll add a comment before resolving this) - we want to leave the parser and its docs open. The websocket times out (currently after 30s) to avoid eating up a free websocket if there is no activity from the user, but we must maintain the state of the currently open documents.

Reply:
Never mind, I see what you mean.